### ISSUE TITLE
SqlCommand.ExecuteNonQuery always throws "non-blocking socket would block" exception

### ISSUE THREAD

**Issue by rjvdboon** [mono/mono#6350]
## Steps to Reproduce

1. Create a Sql Server database and the sample table (see the CreateTable method in the source)
2. Adapt the sample code below: change the connection string to contain the right database server, database name, user id and password
3. run with mono 5.9+
4. observe the error

## Current Behavior

With 5.9+ it throws an exception on reading from the socket
With 5.8 and before it succeeds.

## Expected Behavior

I expect no exception

### On which platforms did you notice this

[ ] macOS
[ x ] Linux Centos 6.8, RHEL 6.8 and 6.9. Note that it works correcly on Centos 7.3.
[ ] Windows
Centos 6.8 (azure cloud image "CentOS-based 6.8 HPC"):
`uname -a`: `Linux [machinename] 2.6.32-642.15.1.el6.x86_64 #1 SMP Fri Feb 24 14:31:22 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
`gcc --version`: `gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18)`

**Version Used**:

The following mono versions show this problem:
- Branch 2017-12
- Master
- build 5.9.0.380 from [centos6-nightly](http://download.mono-project.com/repo/centos6-nightly/m/mono-core/)

### Stacktrace

```
System.Data.SqlClient.SqlException (0x80131904): Snix_Read (provider: SNI_PN7, error: 35 - SNI_ERROR_35)
Snix_Read (provider: SNI_PN7, error: 35 - SNI_ERROR_35)  ---> System.IO.IOException: Unable to read data from the transport connection: Operation on non-blocking socket would block.  ---> System.Net.Sockets.SocketException: Operation on non-blocking socket would block
  at System.Net.Sockets.Socket.Receive (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.Net.Sockets.SocketFlags socketFlags) [0x00017] in /builddir/build/BUILD/mono-5.9.0.380/mcs/class/referencesource/System/net/System/Net/Sockets/Socket.cs:1773 
  at System.Net.Sockets.NetworkStream.Read (System.Byte[] buffer, System.Int32 offset, System.Int32 size) [0x0009b] in /builddir/build/BUILD/mono-5.9.0.380/mcs/class/referencesource/System/net/System/Net/Sockets/NetworkStream.cs:513 
   --- End of inner exception stack trace ---
  at System.Net.Sockets.NetworkStream.Read (System.Byte[] buffer, System.Int32 offset, System.Int32 size) [0x000c3] in /builddir/build/BUILD/mono-5.9.0.380/mcs/class/referencesource/System/net/System/Net/Sockets/NetworkStream.cs:525 
  at System.Data.SqlClient.SNI.SNIPacket.ReadFromStream (System.IO.Stream stream) [0x00000] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs:280 
  at System.Data.SqlClient.SNI.SNITCPHandle.Receive (System.Data.SqlClient.SNI.SNIPacket& packet, System.Int32 timeoutInMilliseconds) [0x00062] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs:440 
  at System.Data.SqlClient.SqlConnection.OnError (System.Data.SqlClient.SqlException exception, System.Boolean breakConnection, System.Action`1[T] wrapCloseInAction) [0x0005a] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs:1199 
  at (wrapper remoting-invoke-with-check) System.Data.SqlClient.SqlConnection.OnError(System.Data.SqlClient.SqlException,bool,System.Action`1<System.Action>)
  at System.Data.SqlClient.SqlInternalConnection.OnError (System.Data.SqlClient.SqlException exception, System.Boolean breakConnection, System.Action`1[T] wrapCloseInAction) [0x00013] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnection.cs:612 
  at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning (System.Data.SqlClient.TdsParserStateObject stateObj, System.Boolean callerHasConnectionLock, System.Boolean asyncClose) [0x00152] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs:1107 
  at System.Data.SqlClient.TdsParserStateObject.ThrowExceptionAndWarning (System.Boolean callerHasConnectionLock, System.Boolean asyncClose) [0x00000] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs:932 
  at System.Data.SqlClient.TdsParserStateObject.ReadSniError (System.Data.SqlClient.TdsParserStateObject stateObj, System.UInt32 error) [0x00170] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs:2572 
  at System.Data.SqlClient.TdsParserStateObject.ReadSniSyncOverAsync () [0x00073] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs:2107 
  at System.Data.SqlClient.TdsParserStateObject.TryReadNetworkPacket () [0x0003a] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs:2032 
  at System.Data.SqlClient.TdsParserStateObject.TryPrepareBuffer () [0x00049] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs:1077 
  at System.Data.SqlClient.TdsParserStateObject.TryReadByte (System.Byte& value) [0x00019] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs:1311 
  at System.Data.SqlClient.TdsParser.TryRun (System.Data.SqlClient.RunBehavior runBehavior, System.Data.SqlClient.SqlCommand cmdHandler, System.Data.SqlClient.SqlDataReader dataStream, System.Data.SqlClient.BulkCopySimpleResultSet bulkCopyHandler, System.Data.SqlClient.TdsParserStateObject stateObj, System.Boolean& dataReady) [0x0012f] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs:1590 
  at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds (System.String methodName, System.Boolean async, System.Int32 timeout, System.Boolean asyncWrite) [0x0020f] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs:2361 
  at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery (System.Threading.Tasks.TaskCompletionSource`1[TResult] completion, System.Boolean sendToPipe, System.Int32 timeout, System.Boolean asyncWrite, System.String methodName) [0x0006a] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs:1157 
  at System.Data.SqlClient.SqlCommand.ExecuteNonQuery () [0x00049] in /builddir/build/BUILD/mono-5.9.0.380/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs:874 
  at ConsoleApplication1.Program.DoSomeStuff (System.Data.SqlClient.SqlConnection con, System.String query) [0x00008] in <f2f7a830e7fd43759eb874cb5043edbc>:0 
ClientConnectionId:33fbcecb-f673-45c1-bae8-fb1a2db98405 
```

### Sample program code:
```
using System;
using System.Data.SqlClient;

namespace ConsoleApplication1 {
   internal class Program {
      static string connectionString = "Server=dbserver,1433;Database=MonoTestSql;user id=[sa];password=[password]";
      private static void Main(string[] args){
            if (args == null || args.Length == 0)
                 WithDisposesWithClearPools();
            else CreateTable();
      }
      private static void CreateTable(){
         using (var con = new SqlConnection(connectionString)) {
            con.Open();
            DoSomeStuff(con, "CREATE TABLE Dogs1 ( Weight int, Name nvarchar(255))");
         }         
      }

      private static void WithDisposesWithClearPools(){
         Console.WriteLine("WithDisposesWithClearPools");
         SqlConnection con = new SqlConnection(connectionString);

         con.Open();
         DoSomeStuff(con);
         con.Close();
         SqlConnection.ClearAllPools();
      }
      private static void DoSomeStuff(SqlConnection con, string query = "UPDATE Dogs1 SET Weight=1 WHERE Weight < 0") {
         try {
            using (var command = new SqlCommand(query, con)) {
               command.ExecuteNonQuery();
            }
         } catch (Exception ex) {
            Console.WriteLine("exception on {0}: {1}", query, ex.ToString());
         }
      }
   }
}
```

**Comment by marek-safar** [comment]
@EgorBo this is a regression in our port could you investigate

**Comment by rjvdboon** [comment]
Looking into this some more I've got the following observations:
- The exception is during handling of the timeout while waiting for the query result,
- according to Sql Profiler Sql Server didn't even receive a query, so is not sending any result either
![mono_sql5_10_no_request](https://user-images.githubusercontent.com/1241158/34607647-dda4cc84-f214-11e7-8c82-deff81bb52ba.png)

executing `mono --debug --trace=N:System.Data.SqlClient,N:System.Net,N:System.Net.Sockets,N:System.Data.SqlClient.SNI testSqlDispose.exe` gave me the following traces:
Failing: mono 2017-12 Head: [mono510_sqlReadTimeout.txt](https://github.com/mono/mono/files/1606477/mono510_sqlReadTimeout.txt)
OK: mono 5.8.0.88: [mono58088_sql_noProblem.txt](https://github.com/mono/mono/files/1606479/mono58088_sql_noProblem.txt)

**Comment by rjvdboon** [comment]
Is there a sqlclient integration testsuite I can run? And if so, how? Because I think it is not part of a usual `make check`.

**Comment by EgorBo** [comment]
@marek-safar @rjvdboon I can reproduce it on my fedora with mono 5.9 but it works fine with the latest mono 5.11 (master). Looks more like the SSLStream bug I fixed (by a workaround) a while ago for appleTls platforms.

@rjvdboon there is an integration suit:
cd mcs/class/System.Data
```
export SYSTEM_DATA_MSSQL_V3=""Server=dbserver,1433;Database=master;User Id=sa;Password=qwerty123;""
make run-test
```

**Comment by rjvdboon** [comment]
@EgorBo Your comment on SSLStream bug fix made me check something: With GCC 4.4.7 (the default on rhel6/centos6) the BTLS part doesn't compile, and thus falls back to the old mono ssl implementation. If I `configure --disable-btls` on Centos7.3 (GCC 4.8.5), thus using the old mono ssl implementation, the test program fails in the same way.

`make run-test` never finishes as it times out trying to setup the sql database, the test suite is hitting the same problem as my test program.

**Comment by EgorBo** [comment]
@rjvdboon could you please check your test with the following option in the connection string:
`;Encrypt=True;TrustServerCertificate=True;`

**Comment by rjvdboon** [comment]
@EgorBo With that added to the connectionstring it works.

**Comment by EgorBo** [comment]
@rjvdboon thank you for confirming that!

**Comment by rjvdboon** [comment]
Shouldn't this stay open until https://github.com/mono/corefx/pull/14 is applied on mono/corefx, and mono's corefx submodule is bumped?

**Comment by rjvdboon** [comment]
@EgorBo Do you know the magic git incantation to test the PR on corefx in combination with mono? Apparently the PR on corefx, and bump of the mono external, is awaiting something, but I'd like to test in my environment to make sure that it indeed fixes my problem. (With my apologies if this is a newbie question or inappropriate, I'm trying to learn my way around linux, git and the mono build setup...)

**Comment by rjvdboon** [comment]
Please reopen, the issue persists even with these changes.

I've ildasm-ed the System.Data, and it still does the `_sslStream.Dispose()` in `SNITcpHandle.DisableSSL`. 
Analyzing the build log I see a `-d:MONO_FEATURE_APPLETLS` on the build command lines, which seems to be added because `MONO_FEATURE_APPLETLS` is unconditionally set in the `mcs/build/profiles/net_4_x.make` file.

And when I look at the ` mcs/class/corlib/System.Security.Cryptography.X509Certificates/X509Helper.cs` file it seems that `MONO_FEATURE_APPLETLS` by itself only suggests that AppleTLS might be in use, but extra checks are needed to be certain. Maybe the same check should be incorporated in `SNITcpHandle`.

**Comment by EgorBo** [comment]
@rjvdboon oh indeed, good catch, thanks

**Comment by rjvdboon** [comment]
Is see it is updated on mono/corefx@9a6ece72a224d6448ee19cd701a4da50267d313d, I'll check again when mono's corefx external is updated.

**Comment by rjvdboon** [comment]
@marek-safar @EgorBo Shouldn't this issue stay open until the fix is in the mono tree?

**Comment by marek-safar** [comment]
@rjvdboon we usually close the issue once it's merged in master and track any backports elsewhere

**Comment by rjvdboon** [comment]
@marek-safar I agree with that policy, that's how I would do it as well. My point is that the fix is not merged in mono/master yet (it is in corefx/master).
Or do you consider bumping mono/mono/external/corefx a backport? In that case: please leave it closed, I'll check the externals regularly from now on.

**Comment by rjvdboon** [comment]
Verified fixed with commit 1dd3fd1c64406a11ee70a23e8b145d82d31cd950

**Comment by yorb1990** [comment]
SAME ERROR AFTER UPDATE "MONO RUNTIME"
```
MSG:Snix_Read (provider: SNI_PN7, error: 35 - SNI_ERROR_35)
Snix_Read (provider: SNI_PN7, error: 35 - SNI_ERROR_35)
STACK:   at System.Data.SqlClient.SqlConnection.OnError (System.Data.SqlClient.SqlException exception, System.Boolean breakConnection, System.Action`1[T] wrapCloseInAction) [0x0004c] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at (wrapper remoting-invoke-with-check) System.Data.SqlClient.SqlConnection.OnError(System.Data.SqlClient.SqlException,bool,System.Action`1<System.Action>)
  at System.Data.SqlClient.SqlInternalConnection.OnError (System.Data.SqlClient.SqlException exception, System.Boolean breakConnection, System.Action`1[T] wrapCloseInAction) [0x00013] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning (System.Data.SqlClient.TdsParserStateObject stateObj, System.Boolean callerHasConnectionLock, System.Boolean asyncClose) [0x00152] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParserStateObject.ThrowExceptionAndWarning (System.Boolean callerHasConnectionLock, System.Boolean asyncClose) [0x00000] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParserStateObject.ReadSniError (System.Data.SqlClient.TdsParserStateObject stateObj, System.UInt32 error) [0x00165] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParserStateObject.ReadSniSyncOverAsync () [0x00072] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParserStateObject.TryReadNetworkPacket () [0x0003a] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParserStateObject.TryPrepareBuffer () [0x00049] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParserStateObject.TryReadByte (System.Byte& value) [0x00019] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.TdsParser.TryRun (System.Data.SqlClient.RunBehavior runBehavior, System.Data.SqlClient.SqlCommand cmdHandler, System.Data.SqlClient.SqlDataReader dataStream, System.Data.SqlClient.BulkCopySimpleResultSet bulkCopyHandler, System.Data.SqlClient.TdsParserStateObject stateObj, System.Boolean& dataReady) [0x0012f] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds (System.String methodName, System.Boolean async, System.Int32 timeout, System.Boolean asyncWrite) [0x0020f] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery (System.Threading.Tasks.TaskCompletionSource`1[TResult] completion, System.Boolean sendToPipe, System.Int32 timeout, System.Boolean asyncWrite, System.String methodName) [0x0006a] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at System.Data.SqlClient.SqlCommand.ExecuteNonQuery () [0x00049] in <10fa2b0e0d09421cafd1a18c9a010576>:0 
  at bocudas.Index.indexer.TryBuildDocument (System.Collections.Generic.KeyValuePair`2[TKey,TValue] dbs, Lucene.Net.Index.IndexWriter IW, System.Int16 point) [0x00417] in /home/yorb1990/Documentos/VSonline/FGC/ServicioCsharp/bocudas.Index/indexer.cs:547 
SOURCE: Core .Net SqlClient Data Provider

```
